# Eat Safe, Love

## Notebook Set Up

In [59]:
from pymongo import MongoClient
import pandas as pd
from pprint import pprint

In [60]:
# Create an instance of MongoClient
mongo = MongoClient(port=27017)

In [61]:
# assign the uk_food database to a variable name
uk_food = mongo['uk_food']

In [62]:
# review the collections in our database
collections = uk_food.list_collection_names()
print(collections)

['establishments']


In [63]:
# assign the collection to a variable
establishment_data = uk_food['establishments']
establishments_data = establishment_data.find_one()
pprint(establishment_data)




Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'uk_food'), 'establishments')


## Part 3: Exploratory Analysis
Unless otherwise stated, for each question: 
* Use `count_documents` to display the number of documents contained in the result.
* Display the first document in the results using `pprint`.
* Convert the result to a Pandas DataFrame, print the number of rows in the DataFrame, and display the first 10 rows.

### 1. Which establishments have a hygiene score equal to 20?

In [64]:
# Find the establishments with a hygiene score of 20
query = {"scores.Hygiene": 20}

# Use count_documents to display the number of documents in the result
print(establishment_data.count_documents(query))

# Display the first document in the results using pprint
pprint(establishment_data.find_one(query))


41
{'AddressLine1': '5-6 Southfields Road',
 'AddressLine2': 'Eastbourne',
 'AddressLine3': 'East Sussex',
 'AddressLine4': '',
 'BusinessName': 'The Chase Rest Home',
 'BusinessType': 'Caring Premises',
 'BusinessTypeID': 5,
 'ChangesByServerID': 0,
 'Distance': 4613.888288172291,
 'FHRSID': 110681,
 'LocalAuthorityBusinessID': '4029',
 'LocalAuthorityCode': '102',
 'LocalAuthorityEmailAddress': 'Customerfirst@eastbourne.gov.uk',
 'LocalAuthorityName': 'Eastbourne',
 'LocalAuthorityWebSite': 'http://www.eastbourne.gov.uk/foodratings',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': 'BN21 1BU',
 'RatingDate': '2021-09-23T00:00:00',
 'RatingKey': 'fhrs_0_en-gb',
 'RatingValue': 0.0,
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('67a940be0b25680b84144e6c'),
 'geocode': {'latitude': 50.769705, 'longitude': 0.27694},
 'links': [{'href': 'https://api.ratings.food.gov.uk/establishments/110681',
            'rel': 'self'}],
 'meta': {'dataSource': None,
          'extrac

## There are `41` establishments with a hygiene score of 20

The first is `The Chase Rest Home`.


In [65]:
# Convert the result to a Pandas DataFrame
hygiene_20 = pd.DataFrame(list(establishments.find(query)))

# Display the number of rows in the DataFrame
hygiene_count = len(hygiene_20)
print(hygiene_count)

# Display the first 10 rows of the DataFrame
hygiene_20.head(10)


41


_id   FHRSID  ChangesByServerID  \
0  67a940be0b25680b84144e6c   110681                  0   
1  67a940be0b25680b841451f2   612039                  0   
2  67a940be0b25680b841454ff   730933                  0   
3  67a940be0b25680b841456ed   172735                  0   
4  67a940be0b25680b841456fb   172953                  0   
5  67a940bf0b25680b841460b0   512854                  0   
6  67a940bf0b25680b841462b7  1537089                  0   
7  67a940bf0b25680b841477e4   155648                  0   
8  67a940bf0b25680b84147c2b  1012883                  0   
9  67a940bf0b25680b84148438   644109                  0   

  LocalAuthorityBusinessID               BusinessName  \
0                     4029        The Chase Rest Home   
1                1970/FOOD                 Brenalwood   
2                1698/FOOD              Melrose Hotel   
3             PI/000023858              Seaford Pizza   
4             PI/000024532              Golden Palace   
5            12/00816/BUTH           Ashby's Butchers   
6          22/00224/RESTUN  South Sea Express Cuisine   
7              EH/00006058              Golden Palace   
8         17/00110/MIXED/S             The Tulip Tree   
9                    41811                      F & S   

                        BusinessType  BusinessTypeID           AddressLine1  \
0                    Caring Premises               5   5-6 Southfields Road   
1                    Caring Premises               5              Hall Lane   
2  Hotel/bed & breakfast/guest house            7842  53 Marine Parade East   
3             Takeaway/sandwich shop            7844          4 High Street   
4            Restaurant/Cafe/Canteen               1         5 South Street   
5                  Retailers - other            4613   777 Southchurch Road   
6            Restaurant/Cafe/Canteen               1    33 Alexandra Street   
7             Takeaway/sandwich shop            7844          7 London Road   
8            Restaurant/Cafe/Canteen               1          3 The Village   
9                  Retailers - other            4613                          

         AddressLine2 AddressLine3  ...  \
0          Eastbourne  East Sussex  ...   
1  Walton-on-the-Naze        Essex  ...   
2      Clacton On Sea        Essex  ...   
3             Seaford  East Sussex  ...   
4             Seaford  East Sussex  ...   
5     Southend-On-Sea        Essex  ...   
6     Southend-on-sea        Essex  ...   
7            Rayleigh        Essex  ...   
8       Chiddingstone         KENT  ...   
9      81 Southernhay     Basildon  ...   

                      LocalAuthorityWebSite  \
0  http://www.eastbourne.gov.uk/foodratings   
1             http://www.tendringdc.gov.uk/   
2             http://www.tendringdc.gov.uk/   
3       http://www.lewes-eastbourne.gov.uk/   
4       http://www.lewes-eastbourne.gov.uk/   
5                http://www.southend.gov.uk   
6                http://www.southend.gov.uk   
7                http://www.rochford.gov.uk   
8              http://www.sevenoaks.gov.uk/   
9                http://www.basildon.gov.uk   

              LocalAuthorityEmailAddress  \
0        Customerfirst@eastbourne.gov.uk   
1             fhsadmin@tendringdc.gov.uk   
2             fhsadmin@tendringdc.gov.uk   
3    ehealth.ldc@lewes-eastbourne.gov.uk   
4    ehealth.ldc@lewes-eastbourne.gov.uk   
5    EnvironmentalHealth@southend.gov.uk   
6    EnvironmentalHealth@southend.gov.uk   
7       customerservices@rochford.gov.uk   
8  environmental.health@sevenoaks.gov.uk   
9                    ehs@basildon.gov.uk   

                                              scores  SchemeType  \
0  {'Hygiene': 20, 'Structural': 20, 'ConfidenceI...        FHRS   
1  {'Hygiene': 20, 'Structural': 15, 'ConfidenceI...        FHRS   
2  {'Hygiene': 20, 'Structural': 20, 'ConfidenceI...        FHRS   
3  {'Hygiene': 20, 'Structural': 10, 'ConfidenceI...        FHRS   
4  {'Hygiene': 20, 'Structural': 10, 'ConfidenceI...       

### 2. Which establishments in London have a `RatingValue` greater than or equal to 4?

In [66]:
# Find the establishments with London as the Local Authority and has a RatingValue greater than or equal to 4.
query = {
    "AddressLine3": "London",
    "RatingValue": {"$gte": 4}  # Filter for RatingValue >= 4
}

# Use count_documents to display the number of documents in the result
print(establishment_data.count_documents(query))

# Display the first document in the results using pprint
pprint(establishment_data.find_one(query))


1004
{'AddressLine1': '75 Woolwich Road',
 'AddressLine2': 'Abbey Wood',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'The Croft Day Nursery',
 'BusinessType': 'Caring Premises',
 'BusinessTypeID': 5,
 'ChangesByServerID': 0,
 'Distance': 4645.185333904148,
 'FHRSID': 329113,
 'LocalAuthorityBusinessID': '03260/0075/0/000',
 'LocalAuthorityCode': '503',
 'LocalAuthorityEmailAddress': 'food.safety@bexley.gov.uk',
 'LocalAuthorityName': 'Bexley',
 'LocalAuthorityWebSite': 'http://www.bexley.gov.uk',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': 'SE2 0DY',
 'RatingDate': '2019-09-02T00:00:00',
 'RatingKey': 'fhrs_5_en-gb',
 'RatingValue': 5.0,
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('67a940bf0b25680b84149e4c'),
 'geocode': {'latitude': 51.482981, 'longitude': 0.136519},
 'links': [{'href': 'http://api.ratings.food.gov.uk/establishments/329113',
            'rel': 'self'}],
 'meta': {'dataSource': None,
          'extractDate': '0001-01-01T

## There are `1004` London establishments with a RatingValue of greater than or equal to 4

### The first is the `Croft Day Nursery`


In [67]:
# Convert the result to a Pandas DataFrame
london_4ormore = pd.DataFrame(list(establishments.find(query)))

# Display the number of rows in the DataFrame
london_4ormore_count = len(london_4ormore)
print(london_4ormore_count)

# Display the first 10 rows of the DataFrame
london_4ormore.head(10)

1004


_id   FHRSID  ChangesByServerID  \
0  67a940bf0b25680b84149e4c   329113                  0   
1  67a940bf0b25680b84149e4d   329114                  0   
2  67a940bf0b25680b84149fac  1402880                  0   
3  67a940bf0b25680b8414a056   975472                  0   
4  67a940bf0b25680b8414a057  1262185                  0   
5  67a940bf0b25680b8414a1c0   687040                  0   
6  67a940bf0b25680b8414a1d0   610394                  0   
7  67a940bf0b25680b8414a22f  1109051                  0   
8  67a940bf0b25680b8414a252  1385124                  0   
9  67a940bf0b25680b8414a2f5  1256695                  0   

  LocalAuthorityBusinessID                                       BusinessName  \
0         03260/0075/0/000                              The Croft Day Nursery   
1         03260/2001/2/000                     Belvedere Sports & Social Club   
2              21/00281/CP  Independent Catering Management At Dulverton P...   
3              17/00162/CP                                  Abbey Good Coffee   
4              20/00155/CP                            Chestnuts Kiosk Limited   
5         03190/0014/0/000                                    Brownes Chemist   
6         03190/0008/0/000                                       A W Foodhall   
7              19/00026/CP                      BP Thamesmead Service Station   
8                    14474                    Alexander McLeod Primary School   
9              20/00134/CP                                   Lakeside Nursery   

                BusinessType  BusinessTypeID        AddressLine1  \
0            Caring Premises               5    75 Woolwich Road   
1          Pub/bar/nightclub            7843  101A Woolwich Road   
2  School/college/university            7845    Dulverton School   
3    Restaurant/Cafe/Canteen               1  Lesnes Abbey Lodge   
4    Restaurant/Cafe/Canteen               1  Lesnes Abbey Lodge   
5          Retailers - other            4613   14-16 Wilton Road   
6          Retailers - other            4613       8 Wilton Road   
7          Retailers - other            4613     Service Station   
8  School/college/university            7845      Fuchsia Street   
9            Caring Premises               5    Lakeside Complex   

       AddressLine2 AddressLine3  ...             LocalAuthorityWebSite  \
0        Abbey Wood       London  ...          http://www.bexley.gov.uk   
1        Abbey Wood       London  ...          http://www.bexley.gov.uk   
2    Dulverton Road       London  ...          http://www.bexley.gov.uk   
3          New Road       London  ...          http://www.bexley.gov.uk   
4          New Road       London  ...          http://www.bexley.gov.uk   
5        Abbey Wood       London  ...          http://www.bexley.gov.uk   
6        Abbey Wood       London  ...          http://www.bexley.gov.uk   
7   Harrow Manorway       London  ...          http://www.bexley.gov.uk   
8                         London  ...  http://www.royalgreenwich.gov.uk   
9  2 Bazalgette Way       London  ...          http://www.bexley.gov.uk   

     LocalAuthorityEmailAddress  \
0     food.safety@bexley.gov.uk   
1     food.safety@bexley.gov.uk   
2     food.safety@bexley.gov.uk   
3     food.safety@bexley.gov.uk   
4     food.safety@bexley.gov.uk   
5     food.safety@bexley.gov.uk   
6     food.safety@bexley.gov.uk   
7     food.safety@bexley.gov.uk   
8  health@royalgreenwich.gov.uk   
9     food.safety@bexley.gov.uk   

                                              scores  SchemeType  \
0  {'Hygiene': 0, 'Structural': 0, 'ConfidenceInM...        FHRS   
1  {'Hygiene': 5, 'Structural': 5, 'ConfidenceInM...        FHRS   
2  {'Hygiene': 0, 'Structural': 0, 'ConfidenceInM...        FHRS   
3  {'Hygiene': 5, 'Structural': 5, 'ConfidenceInM...        FHRS   
4  {'Hygiene': 5, 'Structural': 5, 'ConfidenceInM...        FHRS   
5  {'Hygiene': 0, 'Structural': 0, 'ConfidenceInM...        FHRS   
6  {'Hygiene': 0, 'Structural': 0, 'Confidenc

### 3. What are the top 5 establishments with a `RatingValue` rating value of 5, sorted by lowest hygiene score, nearest to the new restaurant added, "Penang Flavours"?

In [70]:
# Define the query filter
query = {"BusinessName": "Penang Flavours"}

# Define the projection (only return latitude & longitude)
projection = {
    "geocode.latitude": 1,
    "geocode.longitude": 1,
    "_id": 0  # Exclude the _id field
}

# Execute the query and store the result
result = uk_food["establishments"].find_one(query, projection)

# Extract latitude and longitude as floats
if result:
    latitude = float(result.get("geocode", {}).get("latitude", 0))
    longitude = float(result.get("geocode", {}).get("longitude", 0))
else:
    latitude, longitude = None, None

print(f"For Penang Flavours, the latitude is {latitude} and the longitude is {longitude}")



For Penang Flavours, the latitude is 51.490142 and the longitude is 0.08384


In [82]:
# Search within 0.01 degree on either side of the latitude and longitude.
# Rating value must equal 5
# Sort by hygiene score

degree_search = 0.01

query = {
    "geocode.latitude": {"$gte": latitude - degree_search, "$lte": latitude + degree_search},
    "geocode.longitude": {"$gte": longitude - degree_search, "$lte": longitude + degree_search},
    "scores.Hygiene": {"$ne": None}  # Exclude documents where scores.hygiene is None
    
}

# Limit the fields to be returned to make result more readable

projection = {
    "BusinessName": 1,  # Include business name
    "geocode.latitude": 1,
    "geocode.longitude": 1,
    "scores.Hygiene": 1,  # Include hygiene score
    "_id": 0  # Exclude the _id field
}
   
results = list(uk_food["establishments"].find(query, projection).sort("scores.Hygiene", 1).limit(5))

#Print the results
for doc in results:
    print(doc)



{'BusinessName': 'Angels Bakery', 'scores': {'Hygiene': 0}, 'geocode': {'longitude': 0.0922671, 'latitude': 51.4873404}}
{'BusinessName': 'The Gurkha Club', 'scores': {'Hygiene': 0}, 'geocode': {'longitude': 0.0910482, 'latitude': 51.4865168}}
{'BusinessName': 'Plumstead Manor Nursery', 'scores': {'Hygiene': 0}, 'geocode': {'longitude': 0.0859939977526665, 'latitude': 51.481517791748}}
{'BusinessName': 'Iceland', 'scores': {'Hygiene': 0}, 'geocode': {'longitude': 0.0924199968576431, 'latitude': 51.4871482849121}}
{'BusinessName': 'Howe and Co Fish and Chips - Van 17', 'scores': {'Hygiene': 0}, 'geocode': {'longitude': 0.0925370007753372, 'latitude': 51.4875335693359}}


In [83]:
# Convert result to Pandas DataFrame
near_penang_flavours = pd.DataFrame(results)

# Display the DataFrame
near_penang_flavours

BusinessName          scores  \
0                        Angels Bakery  {'Hygiene': 0}   
1                      The Gurkha Club  {'Hygiene': 0}   
2              Plumstead Manor Nursery  {'Hygiene': 0}   
3                              Iceland  {'Hygiene': 0}   
4  Howe and Co Fish and Chips - Van 17  {'Hygiene': 0}   

                                             geocode  
0   {'longitude': 0.0922671, 'latitude': 51.4873404}  
1   {'longitude': 0.0910482, 'latitude': 51.4865168}  
2  {'longitude': 0.0859939977526665, 'latitude': ...  
3  {'longitude': 0.0924199968576431, 'latitude': ...  
4  {'longitude': 0.0925370007753372, 'latitude': ...

### 4. How many establishments in each Local Authority area have a hygiene score of 0?

In [ ]:
# Create a pipeline that:
# 1. Matches establishments with a hygiene score of 0


# 2. Groups the matches by Local Authority
# 3. Sorts the matches from highest to lowest

pipeline = [
    {"$match": {"scores.Hygiene": 0}},  # Filter establishments with hygiene score = 0
    {"$group": {
        "_id": "$LocalAuthorityName",  # Group by LocalAuthorityName
        "count": {"$sum": 1}  # Count establishments
    }},
    {"$sort": {"count": -1}}  # Sort by count in descending order
]

# Execute the aggregation pipeline
results = list(uk_food["establishments"].aggregate(pipeline))

# Print the number of documents in the result
print(f"Number of documents in the result: {len(results)}")

# Print the first 10 results
print("First 10 results:")
for doc in results[:10]:  # Slice to get the first 10 documents
    print(doc)

Number of documents in the result: 55
First 10 results:
{'_id': 'Thanet', 'count': 1130}
{'_id': 'Greenwich', 'count': 882}
{'_id': 'Maidstone', 'count': 713}
{'_id': 'Newham', 'count': 711}
{'_id': 'Swale', 'count': 686}
{'_id': 'Chelmsford', 'count': 680}
{'_id': 'Medway', 'count': 672}
{'_id': 'Bexley', 'count': 607}
{'_id': 'Southend-On-Sea', 'count': 586}
{'_id': 'Tendring', 'count': 542}


In [87]:
# Convert result to Pandas DataFrame
local_authority_hygeiene0 = pd.DataFrame(results)

# Display the number of rows in the DataFrame
local_authority_hygeiene0_count = len(local_authority_hygeiene0)


# Display the first 10 rows of the DataFrame
local_authority_hygeiene0.head(10)


_id  count
0           Thanet   1130
1        Greenwich    882
2        Maidstone    713
3           Newham    711
4            Swale    686
5       Chelmsford    680
6           Medway    672
7           Bexley    607
8  Southend-On-Sea    586
9         Tendring    542